### PR TITLE
Add servant-client-jsaddle

### PR DESCRIPTION
**Pull request by roberth**
Contributions welcome :)

#### To do

- [ ] hlint
- [ ] see if it builds on the CI: no GHCJS dependency!
- [ ] add usage example
- [ ] improve typedarray use in sendXhr

### CODE REVIEW

**Comment by phadej** [comment]
Oh. This went completely under my radar. I'll check this tomorrow.

**Comment by roberth** [comment]
Blip. :)

**Comment by phadej** [comment]
I'll add this to travis first, let's see then.

**Comment by roberth** [comment]
@phadej did you make any progress with travis ci?

**Comment by alpmestan** [comment]
Hello,

Sorry for the delay! Both @phadej and I have been quite busy lately. I think we just need to add this new package to the `cabal.project` file and use [haskell-ci]() to regenerate the travis script. Do you feel like giving it a shot or do you want us to drive this home?

**Comment by vaibhavsagar** [comment]
@alpmestan I tried to do this here: https://github.com/roberth/servant/pull/1

**Comment by alpmestan** [comment]
@vaibhavsagar Your patch seems to be changing GHC versions, not just the list of packages to test?

**Comment by vaibhavsagar** [comment]
@alpmestan I think I misunderstood what you wanted me to do earlier. I've tried again now, can you have another look?

**Comment by vaibhavsagar** [comment]
I accidentally force-pushed to the wrong branch, sorry about that!

**Comment by vaibhavsagar** [comment]
Yes, the diff is against `servant`'s master branch.

**Comment by alpmestan** [comment]
Apparently there's a conflict with our current travis script now... Sorry about that, do you want to look into it?

**Comment by vaibhavsagar** [comment]
Fixed.

**Comment by alpmestan** [comment]
OK, I now see:

```
cabal: Could not resolve dependencies:
[__0] trying: cookbook-basic-auth-0.1 (user goal)
[__1] trying: base-4.12.0.0/installed-4.1... (dependency of
cookbook-basic-auth)
[__2] next goal: servant-client-jsaddle (user goal)
[__2] rejecting: servant-client-jsaddle-0.13 (conflict:
base==4.12.0.0/installed-4.1..., servant-client-jsaddle => base>=4.7 && <4.11)
[__2] fail (backjumping, conflict set: base, servant-client-jsaddle)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: base, cookbook-basic-auth,
servant-client-jsaddle
```

**Comment by phadej** [comment]
`jsaddle` packages don't support recent versions of dependencies,
e.g. no install plan with GHC-8.6: https://matrix.hackage.haskell.org/#/package/jsaddle/0.9.5.0/ghc-8.6.3@1548602028

I had to add: and an install plan with GHC-8.4.4 is found: (didn't try to build though)

-- jsaddle
allow-newer:
  jsaddle:aeson,
  jsaddle:exceptions,
  jsaddle:lens,
  jsaddle-webkit2gtk:aeson,
  jsaddle-webkit2gtk:haskell-gi-base,
  jsaddle-dom:base-compat,
  jsaddle-dom:lens

-- Ubuntu packages:
-- libgirepository1.0-dev
-- libwebkit2gtk-4.0-dev
```

`aeson` is restricted to `<1.3` (1.3.0.0 uploaded at Wed Mar 7 18:38:53 UTC 2018)
`exceptions` is <0.9; and (0.10 uploaded  at Sun Mar 11 15:01:38 UTC 2018)

---

I cannot commit to maintain `servant-client-jsaddle` if `jsaddle` itself isn't maintained. Getting `jsaddle` to track dependencies with less only few month lag is a hard requirement, unfortunately.

I pushed my branch to https://github.com/haskell-servant/servant/compare/wip-servant-client-jsaddle

**Comment by vaibhavsagar** [comment]
@phadej there's a new release of `jsaddle` out now, and it looks like CI is now blocked on the absence of `libwebgtk-4.0-dev`. I had a quick look but couldn't figure out how to install this on Ubuntu 14.04, do you have any advice?

**Comment by vaibhavsagar** [comment]
It looks like the `jsaddle` [`.travis.yml`](https://github.com/ghcjs/jsaddle/blob/master/.travis.yml) gets around this by using 16.04 instead, is that a change you would be open to?

**Comment by vaibhavsagar** [comment]
Looks like the current blocker is `jsaddle-dom`, but we're making progress :smile:

**Comment by vaibhavsagar** [comment]
@phadej this is no longer blocked.

**Comment by vaibhavsagar** [comment]
Running into the 50 minute build limit on free Travis CI builds. Not sure how to proceed from here.

**Comment by alpmestan** [comment]
Do we need to make it a separate job? Or even a separate repo? Those "details" are quite annoying, it's not the first time the build time limit bites us.

**Comment by vaibhavsagar** [comment]
I think this is partly because I also switched the distro to 16.04 from 14.04 so it's building *everything* from scratch. One idea I had was to submit a separate PR for the distro switch and then build off that because then most dependencies should be cached, but that's a temporary fix (e.g. we will still have a build timeout when we start testing with 8.8).

**Comment by alpmestan** [comment]
Restarting the builds should do the trick though, shouldn't it? Everything should be in cache, right?

**Comment by vaibhavsagar** [comment]
No, the cache isn't updated if the build does not complete :smile:.

**Comment by alpmestan** [comment]
_*sigh*_

I guess the separate PR "hack" is worth a shot, unless @phadej has a better idea or we want to go for a dedicated travis job...

**Comment by vaibhavsagar** [comment]
I created a PR at https://github.com/haskell-servant/servant/pull/1119.

**Comment by vaibhavsagar** [comment]
I'm going to try building `servant` without `servant-client-jsaddle` and its dependencies in order to prime the cache, and then include `servant-client-jsaddle`, which will hopefully finish this time.

**Comment by vaibhavsagar** [comment]
Okay, the build is actually failing now. Progress!

**Comment by vaibhavsagar** [comment]
I can't seem to reproduce this build failure locally.

**Comment by vaibhavsagar** [comment]
Ah, it looks like this is due to https://github.com/haskell-servant/servant/commit/4fab471c29db5f8fe18b6718b7569cb209538ef8.

**Comment by vaibhavsagar** [comment]
The 8.4 and 8.6 builds are choking on building `jsaddle-dom`, and I'm beginning to think that replacing `jsaddle-webkit2gtk` with something else like `jsaddle-warp` might reduce the number of dependencies and finally allow this build to complete.

**Comment by alpmestan** [comment]
We have faith in you @vaibhavsagar.

**Comment by phadej** [comment]
Not requiring GUI toolkits is definitely a bonus

File: servant-client-jsaddle/servant-client-jsaddle.cabal [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  build-depends:
+      base                  >= 4.7      && < 4.13
+    , bytestring            >= 0.10     && < 0.11
+    , case-insensitive      >= 1.2.0.0  && < 1.3.0.0
+    , containers            >= 0.5      && < 0.7
+    , exceptions            >= 0.8      && < 0.11
+    , http-media
+    , http-types
+    , jsaddle
+    , jsaddle-dom
```
[the remark this answers is not in the document]
Comment (roberth) [reply servant-client-jsaddle/servant-client-jsaddle.cabal]: Reason was portability. Looks like ghcjs-dom provides the same portability, so sgtm.

[the remark this answers is not in the document]
Comment (vaibhavsagar) [reply servant-client-jsaddle/servant-client-jsaddle.cabal]: @eskimor using `ghcjs-dom` might also unblock the 8.4 and 8.6 builds, which are currently failing when trying to compile `JSDOM.Types`

[the remark this answers is not in the document]
Comment (eskimor) [reply servant-client-jsaddle/servant-client-jsaddle.cabal]: Probably not, as ghcjs-dom has a similar module - and on ghc it is even the same.

[the remark this answers is not in the document]
Comment (eskimor) [reply servant-client-jsaddle/servant-client-jsaddle.cabal]: (ghcjs-dom uses jsaddle-dom on ghc under the hood)

[the remark this answers is not in the document]
Comment (vaibhavsagar) [reply servant-client-jsaddle/servant-client-jsaddle.cabal]: I see.

**Comment by dredozubov** [comment]
I was able to build it with jsaddle-warp and updated 0.16: https://github.com/dredozubov/servant/tree/wip-servant-client-jsaddle-016 I've probably messed up the .travis.yml merge, but overall it's good news

File: servant-client-jsaddle/servant-client-jsaddle.cabal [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  exposed-modules:
+    Servant.Client.JSaddle
+    Servant.Client.Internal.JSaddleXhrClient
+  build-depends:
+      base                  >= 4.7      && < 4.13
+    , bytestring            >= 0.10     && < 0.11
+    , case-insensitive      >= 1.2.0.0  && < 1.3.0.0
+    , containers            >= 0.5      && < 0.7
+    , exceptions            >= 0.8      && < 0.11
+    , http-media
```
Comment (hvr) [inline servant-client-jsaddle/servant-client-jsaddle.cabal]: Why are version bounds ommitted for these deps?

File: servant-client-jsaddle/servant-client-jsaddle.cabal [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  location: http://github.com/haskell-servant/servant.git
+
+library
+  exposed-modules:
+    Servant.Client.JSaddle
+    Servant.Client.Internal.JSaddleXhrClient
+  build-depends:
+      base                  >= 4.7      && < 4.13
+    , bytestring            >= 0.10     && < 0.11
+    , case-insensitive      >= 1.2.0.0  && < 1.3.0.0
```
Comment (hvr) [inline servant-client-jsaddle/servant-client-jsaddle.cabal]:
```suggestion
    , case-insensitive      >= 1.2.0.0  && < 1.3
```

**Comment by hvr** [comment]
Btw, what's blocking this PR currently? Also, it seems like the `.travis.yml` file diverged from the merge base?

**Comment by phadej** [comment]
This version of the patch requires xfvb and libwebkit2gtk-4.0-dev and other crap from CI. I recall there was a branch of this branch which doesn't need all of that, but it didn't emerge into PR.

To be clear: I know very little about jsaddle, so cannot cleanup myself. This PR isn't such I could just press merge.

**Comment by dredozubov** [comment]
I have a branch that I updated to use 0.16 https://github.com/dredozubov/servant/tree/wip-servant-client-jsaddle-016

It has a few rough edges: I may have screwed up the merge of .travis.yml(I haven't bothered to check) and I skipped the streaming part for the request handling. Nonetheless, it builds with jsaddle-warp and I've used it in an application.

**Comment by phadej** [comment]
`.travis.yml` can be regenerated with [`haskell-ci`](https://github.com/haskell-ci/haskell-ci), if it conflicts, just regenerate.

This is @roberths branch, I have no idea who can or cannot push there; but I wouldn't.

**Comment by eskimor** [comment]
@dredozubov just create a PR for the PR  and ping me.

**Comment by phadej** [comment]
Btw, I'd like the clean-history PR at the end. Dozens of commits with merge commits in between won't be merged.

**Comment by phadej** [comment]
`servant-jsaddle` merged.

**Pull request closed without merging**